### PR TITLE
ArtifactRepository tenant aware.

### PR DESCRIPTION
**Pull request by kaizimmerm**
Kind of design flaw. ArtifactRepository is not multi-tenancy ready not in API as well as the three implementations.

Added tenant to the API and to the three implementations:
* file + S3 -> tenant is a root folder
* MongoDb GridFS -> tenant is a metadata field

Reviewers: @michahirsch @schabdo 

Fixes as well a computation bug in the DosFilter

Signed-off-by: kaizimmerm <kai.zimmermann@bosch-si.com>

### CODE REVIEW

File: extensions/hawkbit-extension-artifact-repository-mongo/src/main/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStore.java [hunk 1]
```diff
@@ -143,18 +143,20 @@ private void deleteArtifact(final GridFSDBFile dbFile) {
...
+    private DbArtifact store(final String tenant, final InputStream content, final String contentType,
+            final OutputStream os, final File tempFile, final DbArtifactHash hash) {
         final GridFsArtifact storedArtifact;
         try {
             final String sha1Hash = computeSHA1Hash(content, os, hash != null ? hash.getSha1() : null);
             // upload if it does not exist already, check if file exists, not
             // tenant specific.
-            final GridFSDBFile result = gridFs.findOne(new Query().addCriteria(Criteria.where(FILENAME).is(sha1Hash)));
+            final GridFSDBFile result = gridFs.findOne(new Query()
+                    .addCriteria(Criteria.where(FILENAME).is(sha1Hash).and(TENANT_QUERY).is(tenant.toUpperCase())));
```
Comment (michahirsch) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/main/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStore.java]: In this method the `tenant.toUpperCase()` is called twice. You could make to it upper-case one time and store in a local variable. Because the tenant string needs to be sanitized across the class it's maybe nicer to introduce a private method 
```
public String sanitizeTenant(final String tenant)
{
   return tenant.toUpperCase();
}
```

File: extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java [hunk 2]
```diff
@@ -20,54 +20,76 @@
...
+
+        final String sha1 = BaseEncoding.base16().lowerCase().encode(digestInputStream.getMessageDigest().digest());
+        assertThat(artifactStoreUnderTest.getArtifactBySha1(tenant, sha1)).isNotNull();
+        return sha1;
     }
 
     @Test
-    @Description("Ensures that search by MD5 hash finds the expected results.")
-    public void findArtifactByMD5Hash() throws NoSuchAlgorithmException {
+    @Description("Deltes file from repoeitory identified ny SHA1 hash as filename.")
```
Comment (kaizimmerm) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java]: typos

File: extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java [hunk 3]
```diff
@@ -20,54 +20,76 @@
...
+        final String sha1 = storeAndVerify(TENANT, filelengthBytes, filename, contentType);
+
+        artifactStoreUnderTest.deleteBySha1(TENANT, sha1);
+        assertThat(artifactStoreUnderTest.getArtifactBySha1(TENANT, sha1)).isNull();
+    }
+
+    @Test
+    @Description("Verfies that all data of a tenant is erased if repositoty is asked to do so. "
+            + "Data of other tenants is not affected.")
+    public void deleteTenant() throws NoSuchAlgorithmException {
```
Comment (kaizimmerm) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java]: test delete on non existing tenant.

File: extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java [hunk 4]
```diff
@@ -162,26 +169,34 @@ private ObjectMetadata createObjectMetadata(final String mdMD5Hash16, final Stri
...
-        amazonS3.deleteObject(new DeleteObjectRequest(s3Properties.getBucketName(), sha1Hash));
+    public void deleteBySha1(final String tenant, final String sha1Hash) {
+        final String key = objectKey(tenant, sha1Hash);
+
+        LOG.info("Deleting S3 object from bucket {} and key {}", s3Properties.getBucketName(), key);
+        amazonS3.deleteObject(new DeleteObjectRequest(s3Properties.getBucketName(), key));
+    }
+
+    private static String objectKey(final String tenant, final String sha1Hash) {
+        return tenant.trim().toUpperCase() + "/" + sha1Hash;
```
Comment (michahirsch) [inline extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java]: small one: you could call the `sanitizeTenant()` method which does exactly this `tenant.trim().toUpperCase()`

File: extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java [hunk 5]
```diff
@@ -220,4 +235,21 @@ private static DigestOutputStream openFileOutputStream(final File file, final Me
     private boolean exists(final String sha1) {
         return amazonS3.doesObjectExist(s3Properties.getBucketName(), sha1);
     }
+
+    @Override
+    public void deleteByTenant(final String tenant) {
+        final String folder = tenant.toUpperCase();
```
Comment (michahirsch) [inline extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java]: `tenant.toUpperCase();` here is no `trim()` but in the `sanitizeTenant()` you trim the tenant, should be same behavior.

File: extensions/hawkbit-extension-artifact-repository-mongo/src/main/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStore.java [hunk 6]
```diff
@@ -143,18 +147,21 @@ private void deleteArtifact(final GridFSDBFile dbFile) {
 
     }
 
-    private DbArtifact store(final InputStream content, final String contentType, final OutputStream os,
+    private DbArtifact store(final String t, final InputStream content, final String contentType, final OutputStream os,
```
Comment (schabdo) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/main/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStore.java]: Was `t` as parameter name on purpose or by accident?

File: extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java [hunk 7]
```diff
@@ -20,54 +20,77 @@
...
-    private GridFsOperations gridFs;
-
     @Test
     @Description("Ensures that search by SHA1 hash (which is used by hawkBit as artifact ID) finds the expected results.")
     public void findArtifactBySHA1Hash() throws NoSuchAlgorithmException {
         final int filelengthBytes = 128;
         final String filename = "testfile.json";
         final String contentType = "application/json";
 
+        storeAndVerify(TENANT, filelengthBytes, filename, contentType);
```
Comment (schabdo) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java]: What about adding another test that verifies that storing of artefacts is really tenant specific now?

1. Store file for tenant1 
2. Store file for tenant2
3. Try to read file for tenant1 but use SHA1 hash from tenant2 artifact

File: extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java [hunk 8]
```diff
@@ -20,54 +20,77 @@
...
+    }
+
+    @Step
+    private String storeAndVerify(final String tenant, final int filelengthBytes, final String filename,
```
Comment (schabdo) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java]: Every test case is defining `filelengthBytes`, `filename` and `contentType` with the exact same values and afterwords passing them to this method. May I ask you to remove the parameters and inline them?

Comment (schabdo) [inline extensions/hawkbit-extension-artifact-repository-mongo/src/test/java/org/eclipse/hawkbit/artifact/repository/MongoDBArtifactStoreTest.java]: Maybe than `storeRandomArtefactAndVerify`would be a better name for this method

File: extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java [hunk 9]
```diff
@@ -220,4 +235,25 @@ private static DigestOutputStream openFileOutputStream(final File file, final Me
...
+
+        LOG.info("Deleting S3 object folder (tenant) from bucket {} and key {}", s3Properties.getBucketName(), folder);
+
+        // Delete artifacts
+        ObjectListing objects = amazonS3.listObjects(s3Properties.getBucketName(), folder + "/");
+        do {
+            for (final S3ObjectSummary objectSummary : objects.getObjectSummaries()) {
+                amazonS3.deleteObject(s3Properties.getBucketName(), objectSummary.getKey());
+            }
+            objects = amazonS3.listNextBatchOfObjects(objects);
```
Comment (schabdo) [inline extensions/hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Repository.java]: Don't you have to call `objects.isTruncated()` first before requesting the next bunch of objects?

File: hawkbit-artifact-repository-filesystem/src/test/java/org/eclipse/hawkbit/artifact/repository/ArtifactFilesystemRepositoryTest.java [hunk 10]
```diff
@@ -61,16 +62,26 @@ public void getStoredArtifactBasedOnSHA1Hash() {
...
-        artifactFilesystemRepository.deleteBySha1(artifact.getHashes().getSha1());
+        artifactFilesystemRepository.deleteBySha1(TENANT, artifact.getHashes().getSha1());
 
-        assertThat(artifactFilesystemRepository.getArtifactBySha1(artifact.getHashes().getSha1())).isNull();
+        assertThat(artifactFilesystemRepository.getArtifactBySha1(TENANT, artifact.getHashes().getSha1())).isNull();
+    }
+
+    @Test
+    @Description("Verfies that all artifacts of a tenant can be deleted in the file-system repository")
+    public void deleteStoredArtifactOftenant() {
```
Comment (schabdo) [inline hawkbit-artifact-repository-filesystem/src/test/java/org/eclipse/hawkbit/artifact/repository/ArtifactFilesystemRepositoryTest.java]: CAPITAL "T" => deleteStoredArtifactOfTenant

File: hawkbit-artifact-repository-filesystem/src/test/java/org/eclipse/hawkbit/artifact/repository/ArtifactFilesystemRepositoryTest.java [hunk 11]
```diff
@@ -61,16 +62,26 @@ public void getStoredArtifactBasedOnSHA1Hash() {
...
+
+        assertThat(artifactFilesystemRepository.getArtifactBySha1(TENANT, artifact.getHashes().getSha1())).isNull();
     }
 
     @Test
     @Description("Verfies that an artifact which does not exists is deleted quietly in the file-system repository")
     public void deleteArtifactWhichDoesNotExistsBySHA1HashWithoutException() {
         try {
-            artifactFilesystemRepository.deleteBySha1("sha1HashWhichDoesNotExists");
+            artifactFilesystemRepository.deleteBySha1(TENANT, "sha1HashWhichDoesNotExists");
```
Comment (schabdo) [inline hawkbit-artifact-repository-filesystem/src/test/java/org/eclipse/hawkbit/artifact/repository/ArtifactFilesystemRepositoryTest.java]: May I ask you to add a test with a none existing tenant as well. Shoudn't be a problem but better safe than sorry

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/ControllerManagement.java [hunk 12]
```diff
@@ -118,13 +118,15 @@
     /**
      * Retrieves oldest {@link Action} that is active and assigned to a
      * {@link Target}.
+     * 
+     * For performance reasons this method does not throw
+     * {@link EntityNotFoundException} in case target with given controlelrId
+     * does not exist butt will return an {@link Optional#empty()} instead.
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/ControllerManagement.java]: butt => but

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/ActionRepository.java [hunk 13]
```diff
@@ -117,6 +117,17 @@
     Optional<Action> findFirstByTargetControllerIdAndActive(final Sort sort, final String controllerId, boolean active);
 
     /**
+     * Checks if an active action exist for given
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/ActionRepository.java]: exist => exists
